modules/shared/src/main/scala/org/tessellation/schema/GlobalSnapshotInfo.scala

```diff
@@ -1,20 +1,79 @@
 package org.tessellation.schema
 
+import cats.MonadThrow
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.functor._
+
 import scala.collection.immutable.SortedMap
 
+import org.tessellation.currency.schema.currency.{CurrencyIncrementalSnapshot, CurrencySnapshotInfo}
+import org.tessellation.ext.crypto._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.merkletree.syntax._
+import org.tessellation.merkletree.{MerkleRoot, Proof}
 import org.tessellation.schema.address.Address
 import org.tessellation.schema.balance.Balance
-import org.tessellation.schema.snapshot.SnapshotInfo
+import org.tessellation.schema.snapshot.{SnapshotInfo, StateProof}
 import org.tessellation.schema.transaction.TransactionReference
 import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
 
 import derevo.cats.{eqv, show}
 import derevo.circe.magnolia.{decoder, encoder}
 import derevo.derive
 
 @derive(encoder, decoder, eqv, show)
-case class GlobalSnapshotInfo(
+case class GlobalSnapshotInfoV1(
   lastStateChannelSnapshotHashes: SortedMap[Address, Hash],
   lastTxRefs: SortedMap[Address, TransactionReference],
   balances: SortedMap[Address, Balance]
-) extends SnapshotInfo
+) extends SnapshotInfo[GlobalSnapshotStateProof] {
+  def stateProof[F[_]: MonadThrow: KryoSerializer]: F[GlobalSnapshotStateProof] =
+    GlobalSnapshotInfoV1.toGlobalSnapshotInfo(this).stateProof[F]
+}
+
+object GlobalSnapshotInfoV1 {
+  implicit def toGlobalSnapshotInfo(gsi: GlobalSnapshotInfoV1): GlobalSnapshotInfo =
+    GlobalSnapshotInfo(
+      gsi.lastStateChannelSnapshotHashes,
+      gsi.lastTxRefs,
+      gsi.balances,
+      SortedMap.empty,
+      SortedMap.empty
+    )
+}
+
+@derive(encoder, decoder, eqv, show)
+case class GlobalSnapshotStateProof(
+  lastStateChannelSnapshotHashesProof: Hash,
+  lastTxRefsProof: Hash,
+  balancesProof: Hash,
+  lastCurrencySnapshotsProof: Option[MerkleRoot]
+) extends StateProof
+
+object GlobalSnapshotStateProof {
+  def apply(a: (Hash, Hash, Hash, Option[MerkleRoot])): GlobalSnapshotStateProof =
+    GlobalSnapshotStateProof(a._1, a._2, a._3, a._4)
+}
```

Comment:
Not necessary. You can just use `GlobalSnapshotProof.tupled(x)` instead of `GlobalSnapshotProof.apply(x)`

Reply:
cool. I was looking for that